lib/internal/fs/promises.js

```diff
@@ -739,6 +747,68 @@ async function readFile(path, options) {
   return PromisePrototypeFinally(readFileHandle(fd, options), fd.close);
 }
 
+async function stream(handle, dest, options) {
+  checkAborted(options.signal);
+  if (!EE) {
+    EE = require('events');
+  }
+  const BUFFER_SIZE = 128 * 2 ** 10;
+  const {
+    start = 0,
+    end,
+  } = options;
+
+  if (start && typeof start !== 'number') {
+    throw new ERR_INVALID_ARG_TYPE('start', 'number', start);
+  }
+  if (end && typeof end !== 'number') {
+    throw new ERR_INVALID_ARG_TYPE('end', 'number', end);
+  }
+  if (start < 0) {
+    throw new ERR_OUT_OF_RANGE('start', 'greater than 0', start);
+  }
+  if (end < 0) {
+    throw new ERR_OUT_OF_RANGE('end', 'greater than 0', end);
+  }
+  if (end && start > end) {
+    throw new ERR_OUT_OF_RANGE('start', `less than end ${end}`, start);
+  }
+
+  let currPos = start;
+  const allocatedBuffer = Buffer.allocUnsafe(
+    end ? MathMin(end - currPos, BUFFER_SIZE) : BUFFER_SIZE
+  );
+
+  while (true) {
+    if (end && currPos >= end) {
+      break;
+    }
+
+    checkAborted(options.signal);
+    if (dest.writableNeedDrain) {
+      await EE.once(dest, 'drain');
+    }
+
+    const { buffer, bytesRead } = await handle.read(
+      allocatedBuffer,
+      0,
+      allocatedBuffer.byteLength,
+      currPos
+    );
+
+    if (dest.destroyed) {
+      throw new ERR_STREAM_PREMATURE_CLOSE();
+    }
+
+    if (bytesRead === 0) {
+      break;
+    }
+
```

Comment:
Also you need something like:

```js
if (bytesRead !== buffer.length) {
        // Slow path. Shrink to fit.
        // Copy instead of slice so that we don't retain
        // large backing buffer for small reads.
        const dst = Buffer.allocUnsafeSlow(bytesRead)
        buffer.copy(dst, 0, 0, bytesRead)
        buffer = dst
      }
      ```